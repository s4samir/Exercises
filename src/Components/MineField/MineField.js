/*
 * Authored by Samir Kishore
 */
import React, { useState, useRef, useEffect } from "react";
import PropTypes from "prop-types";
import MineCell from "./MineCell/MineCell";
import "./MineField.scss";

export default function MineField({
  row,
  col,
  mineMap,
  startTimer,
  closeGame,
}) {
  const [gameOver, setGameOver] = useState(false);
  const [gameWon, setGameWon] = useState(false);
  const keyClicked = useRef({});
  const successfulClick = useRef(0);

  const prepareField = (row, col) => {
    const rows = [];
    for (let i = 1; i <= row; i++) {
      const cells = [];
      for (let j = 1; j <= col; j++) {
        const key = `${i}${j}`;
        const mine = gameOver ? isMine(key) : null;
        cells.push(
          <MineCell
            key={key}
            number={key}
            isMine={mine}
            onClick={onCellClick}
          />
        );
      }
      rows.push(
        <div className="minerow" key={i}>
          {cells}
        </div>
      );
    }
    return rows;
  };

  const isMine = (key) => mineMap.has(key);

  const isDone = () => successfulClick.current === row * col - mineMap.size;

  const onCellClick = (key) => {
    if (keyClicked.current[key]) {
      return keyClicked.current[key];
    }
    if (isMine(key)) {
      setGameOver(true);
      startTimer(false);
      keyClicked.current[key] = -1;
    } else {
      startTimer(true);
      keyClicked.current[key] = calculateMineAround(key, row, col, mineMap);
      successfulClick.current++;
      if (isDone()) {
        setGameWon(true);
      }
    }
    return keyClicked.current[key];
  };

  useEffect(() => {
    if (gameOver) {
      closeGame("over");
    }
    if (gameWon) {
      closeGame("won");
    }
  }, [gameOver, gameWon, closeGame]);
  return <div className={"minefield"}>{prepareField(row, col)}</div>;
}

const calculateMineAround = (key, row, col, mineMap) => {
  const neighbour = [
    // 8 neighbours of a node
    [-1, -1],
    [-1, 0],
    [-1, 1],
    [0, -1],
    [0, 1],
    [1, -1],
    [1, 0],
    [1, 1],
  ];
  const [i, j] = key;
  let count = 0;
  for (const curr of neighbour) {
    const [i1, j1] = curr;
    const [nextI, nextJ] = [+i + i1, +j + j1];

    if (
      nextI >= 1 &&
      nextI <= row &&
      nextJ >= 1 &&
      nextJ <= col &&
      mineMap.has(`${nextI}${nextJ}`)
    ) {
      count++;
    }
  }
  return count;
};

MineField.propTypes = {
  row: PropTypes.number,
  col: PropTypes.number,
  mineMap: PropTypes.any,
  startTimer: PropTypes.func,
  closeGame: PropTypes.func,
};

MineField.defaultProps = {
  row: 5,
  col: 5,
  mineMap: new Set(),
  startTimer: () => {},
  closeGame: () => {},
};
