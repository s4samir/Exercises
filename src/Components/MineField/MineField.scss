$blast: radial-gradient(yellow, red);
@mixin row {
  .minerow {
    display: flex;

    .minecell {
      width: 5em;
      height: 5em;
      display: flex;
      place-self: center;
      border: 5px outset lightgrey;
      outline: none;
      &.blast{
        background: $blast;
        border-style: inset;
      }
      &.selected{
        border-style: inset;
      }
    }
  }
}
.minefield {
  @include row;
  display: flex;
  flex-direction: column;
}
