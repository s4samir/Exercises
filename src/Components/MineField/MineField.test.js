/*
 * Authored by Samir Kishore
 */
import React from "react";
import { create } from "react-test-renderer";
import ReactTestUtils from "react-dom/test-utils";
import ReactDOM from "react-dom";
import MineField from "./MineField";

describe("Minefield", () => {
  let container;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  it("creates minecell", () => {
    const props = {
      row: 5,
      col: 5,
      mineMap: new Set(),
      startTimer: () => {},
      closeGame: () => {},
    };
    const wrapper = create(<MineField {...props} />);

    expect(wrapper.toJSON()).toMatchSnapshot();
    const instance = wrapper.root;
    expect(instance.findAllByType("button").length).toBe(25);
  });

  it("triggers game over on mine click", () => {
    let mySet = new Set();
    mySet.add("11");
    mySet.add("22");

    const props = {
      row: 3,
      col: 3,
      mineMap: mySet,
      startTimer: jest.fn(() => {}),
      closeGame: jest.fn(() => {}),
    };
    ReactTestUtils.act(() => {
      ReactDOM.render(<MineField {...props} />, container);
    });
    const buttons = container.querySelectorAll("button");

    ReactTestUtils.act(() => {
      buttons[0].dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });

    expect(props.closeGame).toHaveBeenCalledWith("over");
    expect(props.startTimer).toHaveBeenCalledWith(false);
  });

  it("triggers game won on all correct click", () => {
    let mySet = new Set();
    mySet.add("11");
    mySet.add("22");

    const props = {
      row: 2,
      col: 2,
      mineMap: mySet,
      startTimer: jest.fn(() => {}),
      closeGame: jest.fn(() => {}),
    };
    ReactTestUtils.act(() => {
      ReactDOM.render(<MineField {...props} />, container);
    });
    const buttons = container.querySelectorAll("button");

    ReactTestUtils.act(() => {
      buttons[1].dispatchEvent(new MouseEvent("click", { bubbles: true }));
      buttons[2].dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });

    expect(props.closeGame).toHaveBeenCalledWith("won");
    expect(props.startTimer).toHaveBeenCalledWith(true);
  });
});
