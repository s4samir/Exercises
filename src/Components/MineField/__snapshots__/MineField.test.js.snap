// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Minefield creates minecell 1`] = `
<div
  className="minefield"
>
  <div
    className="minerow"
  >
    <button
      className="minecell 11"
      onClick={[Function]}
    />
    <button
      className="minecell 12"
      onClick={[Function]}
    />
    <button
      className="minecell 13"
      onClick={[Function]}
    />
    <button
      className="minecell 14"
      onClick={[Function]}
    />
    <button
      className="minecell 15"
      onClick={[Function]}
    />
  </div>
  <div
    className="minerow"
  >
    <button
      className="minecell 21"
      onClick={[Function]}
    />
    <button
      className="minecell 22"
      onClick={[Function]}
    />
    <button
      className="minecell 23"
      onClick={[Function]}
    />
    <button
      className="minecell 24"
      onClick={[Function]}
    />
    <button
      className="minecell 25"
      onClick={[Function]}
    />
  </div>
  <div
    className="minerow"
  >
    <button
      className="minecell 31"
      onClick={[Function]}
    />
    <button
      className="minecell 32"
      onClick={[Function]}
    />
    <button
      className="minecell 33"
      onClick={[Function]}
    />
    <button
      className="minecell 34"
      onClick={[Function]}
    />
    <button
      className="minecell 35"
      onClick={[Function]}
    />
  </div>
  <div
    className="minerow"
  >
    <button
      className="minecell 41"
      onClick={[Function]}
    />
    <button
      className="minecell 42"
      onClick={[Function]}
    />
    <button
      className="minecell 43"
      onClick={[Function]}
    />
    <button
      className="minecell 44"
      onClick={[Function]}
    />
    <button
      className="minecell 45"
      onClick={[Function]}
    />
  </div>
  <div
    className="minerow"
  >
    <button
      className="minecell 51"
      onClick={[Function]}
    />
    <button
      className="minecell 52"
      onClick={[Function]}
    />
    <button
      className="minecell 53"
      onClick={[Function]}
    />
    <button
      className="minecell 54"
      onClick={[Function]}
    />
    <button
      className="minecell 55"
      onClick={[Function]}
    />
  </div>
</div>
`;
